### ISSUE TITLE
NFE while constructing the project imports tree

### ISSUE THREAD

**Issue by lifengl** [dotnet/project-system#6675]
In 16.8: the code below seems to add the same file to the tree, which may just need a simple check to deduplicate items?

System.ArgumentException - Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.Add

> An item with the caption "Sdk.props"(ProjectTreeFlags: ProjectImportImplicit FileOnDisk ProjectImport FileSystemEntity) already exists under parent "Sdk.props" with the same sort order as the item being added.

```
["Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.Add(UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.Add(IProjectTree)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<set_ShowAllFiles>g__AddChild|22_9(IProjectTree2,<>c__DisplayClass22_1&)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<set_ShowAllFiles>g__SyncNode|22_8(IReadOnlyList`1,IProjectTree2)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<set_ShowAllFiles>g__SyncNode|22_8(IReadOnlyList`1,IProjectTree2)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<>c__DisplayClass22_0.<set_ShowAllFiles>b__7(IProjectVersionedValue`1,ConfiguredProjectExports,CancellationToken)
>>>Microsoft.VisualStudio.ProjectSystem.ProjectTreeProviderBase.TreeUpdateSubmission.<UpdateTree>d__19.MoveNext()
>>>System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task)
>>>Microsoft.VisualStudio.ProjectSystem.CoreProjectTreeProviderBase.<>c__DisplayClass76_1.<<Initialize>b__2>d.MoveNext()
>>>System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
>>>Microsoft.VisualStudio.ProjectSystem.CoreProjectTreeProviderBase.<>c__DisplayClass78_0.<<SubmitTreeUpdateCoreAsync>g__waitOnTask|0>d.MoveNext()
>>>System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>>>Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception)
>>>Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
>>>Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__action|0()
>>>GuardMethodClass.GuardMethod(Func`1,Func`2,Func`2)"]
```

**Comment by drewnoakes** [comment]
Potentially related to [another NFE](https://msit.powerbi.com/groups/me/reports/378af503-a084-400f-97d8-610c59c207e3/ReportSection14be6cf25d0777aa59ce?filter=PreviewCallStack%2FTypeAndTarget%20eq%20%27System.ArgumentException%20-%20Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.ReplaceChild%27):

> The node to add collides with a node already present. Parameter name: newChild

```
["Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.ReplaceChild(UnattachedProjectTreeNode,UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.<>c__DisplayClass72_0.<Replace>b__0(UnattachedProjectTreeNode,UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.MutateRecursive(Func`3,ImmutableStack`1,UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.MutateRecursive(Func`3,ImmutableStack`1,UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.Mutate(Func`3,UnattachedProjectTreeNode,Boolean)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.UnattachedProjectTreeNode.Replace(UnattachedProjectTreeNode,UnattachedProjectTreeNode,Boolean)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.ApplyChange(UnattachedProjectTreeNode)
>>>Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.SetDisplayOrder(Int32)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<set_ShowAllFiles>g__SyncNode|22_8(IReadOnlyList`1,IProjectTree2)
>>>Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<>c__DisplayClass22_0.<set_ShowAllFiles>b__7(IProjectVersionedValue`1,ConfiguredProjectExports,CancellationToken)
>>>Microsoft.VisualStudio.ProjectSystem.ProjectTreeProviderBase.TreeUpdateSubmission.<UpdateTree>d__19.MoveNext()
>>>System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task)
>>>Microsoft.VisualStudio.ProjectSystem.CoreProjectTreeProviderBase.<>c__DisplayClass76_1.<<Initialize>b__2>d.MoveNext()
>>>System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task)
>>>System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task)
>>>Microsoft.VisualStudio.ProjectSystem.CoreProjectTreeProviderBase.<>c__DisplayClass78_0.<<SubmitTreeUpdateCoreAsync>g__waitOnTask|0>d.MoveNext()
>>>System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
>>>Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception)
>>>Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
>>>Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__action|0()
>>>GuardMethodClass.GuardMethod(Func`1,Func`2,Func`2)"]
```

**Comment by drewnoakes** [comment]
We have two feedback tickets for this:

1. AB#1202408
2. AB#1249790

The first gives repo https://github.com/microsoft/partsunlimited as a repro, specifically the web app project. However testing this locally does not reproduce the issue, and the tree appears correctly with duplicate captions:

![image](https://user-images.githubusercontent.com/350947/104171418-bb3a1a80-5456-11eb-9450-1010a3f451fd.png)

I've asked the customers who reported these feedback tickets for clarification on repro'ing the issue.

**Comment by drewnoakes** [comment]
The problem seems to be when we have an existing tree, then change it. If the display order of an items changes, we update it. However if another item in the tree with the same caption already has this display order value, then we fail.

```
Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.ApplyChange(UnattachedProjectTreeNode)
Microsoft.VisualStudio.ProjectSystem.Designers.ProjectTree.SetDisplayOrder(Int32)
Microsoft.VisualStudio.ProjectSystem.Tree.ProjectImports.ImportTreeProvider.<set_ShowAllFiles>g__SyncNode|22_8(IReadOnlyList`1,IProjectTree2)
```

**Comment by drewnoakes** [comment]
I've found a repro.

```
Project.csproj
MyFile.props
MyFolder\MyFile.props
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>
  <Import Project="MyFolder\MyFile.props" />
  <Import Project="MyFile.props" />
  <Import Project="MyFolder\MyFile.props" />
</Project>
```

Open this project. Move the first `Import` to the end of the file and save.

> The node to add collides with a node already present.

**Comment by lifengl** [comment]
ok, there are many problems in the import tree code:
   tree = ReplaceChild(child, child.SetDisplayOrder(displayOrder));

1, it can be changed to tree = child.SetDisplayOrder(displayOrder).Root.  Changing a property of a child node returns an entire new tree (and all parents are updated), there is no reason to do extra ReplaceChild.

2, ReplaceChild implementation only works for a single flat tree, which might be copied from dependencies tree? 
it need call parent.RemoveChild(), but cannot use root.RemoveChild(), so the code should be
 oldChild.Parent.RemoveChild(oldChild).Add(newChild).Root

3, using a collection like existingChildByPath to get a collection of tree node first, then apply change in a loop is invalid code.  Because it captures tree node from the original tree state, after the first change, it should no longer uses tree node from an old state to modify the tree further. The code must find tree node after every change again, and cannot use an earlier cached state. Basically, it can only cache Ids of node, and Find the tree node to be changed right before each change, and it must be repeated after every change.

3,

**Comment by drewnoakes** [comment]
Thank you for looking over this code @lifengl. Some comments and discussion.

1. This optimisation makes sense. I'll submit a PR with this and some other tidy up soon.
2. I think I see what you're saying, and I don't think it applies here. We recursively walk down the tree, building (or updating) it bottom-up. That means we cannot use `.Root` but must use `.Parent` or equivalent. I've looked through the implementation of `ReplaceChild` and think it's fine for our purposes.
3. This makes sense. I've changed the loop to look up the relevant item on each iteration. This changes it from O(N) to O(N^2), but N is usually a small number (2-10) so I don't believe this will be problematic. If we see this on traces, we can investigate further.

As for the bug here, the fix is actually more tricky than the above. In the repro I give above there are two items with identical captions but different file paths, and we see a bug when we transpose them. This means their display order values need to be swapped. Imagine they have orders 1 and 2. The update wants to change 1 to 2, but there's already an item matching 2. The `ReplaceChild` call successfully removes 1, but fails on inserting 2 as the item already exists. It never gets as far as changing the display order on the other item from 2 to 1.

The CPS tree uses comparator `DisplayOrderProjectTreeSort` (which derives from `ProjectTreeSort`) to order a tree node's `Children` collection. This comparator considers the `DisplayOrder` and `Caption` properties (and a few other things). A potential fix here may be to have that comparator also consider `FilePath` if all other values are equal, to break the tie. I don't know what this might break on the CPS side.

Another fix would be for the sync to remove all children before rebuilding, so that there is no moment in time during which there are multiple identical children. This may be wasteful, but updates to this tree are very infrequent.

Another fix would be to prevent duplicate captions somehow, potentially by appending some kind of suffix. This seems the best approach so I will investigate it first.
